pkg/service/backup/worker_deduplicate.go

```diff
@@ -0,0 +1,194 @@
+	for id, ssTableContent := range ssTablesGroupByID {
+		// Check if id is an integer (no UUID)
+		if _, err := strconv.Atoi(id); err == nil {
+			continue
+		}
+		// Check if SSTable is available on remote already
+		if _, ok := remoteSSTables[id]; !ok {
+			continue
+		}
+		// Remove duplicated SSTable from local snapshot
+		for _, file := range ssTableContent {
+			// remove if exists in remote
+			localSSTableFileNameWithPath := path.Join(d.Path, file)
+			w.Logger.Debug(ctx, "Removing local snapshot file (deduplication based on generation UUID)",
+				"host", h.IP, "file", localSSTableFileNameWithPath)
+			info, err := w.Client.RcloneFileInfo(ctx, h.IP, localSSTableFileNameWithPath)
+			if err != nil {
+				return deduplicated, errors.Wrapf(err, "get file info %s", localSSTableFileNameWithPath)
+			}
+			if err := w.Client.RcloneDeleteFile(ctx, h.IP, localSSTableFileNameWithPath); err != nil {
+				return deduplicated, errors.Wrapf(err, "cannot delete local snapshot's SSTable file %s", localSSTableFileNameWithPath)
+			}
+			deduplicated += info.Size
+		}
+	}
+
+	return deduplicated, nil
+}
+
+func (w *worker) basedOnCrc32Content(ctx context.Context, d snapshotDir, h hostInfo, dataDst string,
+	remoteSSTables map[string]struct{}, ssTablesGroupByID map[string][]string,
+) (deduplicated int64, err error) {
+	// Reference to SSTables 3.0 Data File Format
+	// https://opensource.docs.scylladb.com/stable/architecture/sstable/sstable3/sstables-3-data-file-format.html
+
+	// Per every SSTable files group, compare local <ID>-Digest.crc32 content
+	// to the remote <ID>-Digest.crc32 content.
+	// The same content implies that SSTable can be deduplicated and removed from local directory.
+	for id, ssTableContent := range ssTablesGroupByID {
+		crc32Idx := slices.IndexFunc(ssTableContent, func(s string) bool {
+			return strings.HasSuffix(s, "Digest.crc32")
+		})
+		if crc32Idx == -1 {
+			continue
+		}
+		if _, ok := remoteSSTables[id]; !ok {
+			continue
+		}
+
+		remoteCRC32FileNameWithPath := path.Join(dataDst, ssTableContent[crc32Idx])
+		remoteCRC32, err := w.Client.RcloneCat(ctx, h.IP, remoteCRC32FileNameWithPath)
+		if err != nil {
+			if strings.Contains(err.Error(), "object not found") {
+				continue
+			}
+			return deduplicated, errors.Wrapf(err, "cannot get content of remote CRC32 %s", remoteCRC32FileNameWithPath)
+		}
+
+		localCRC32FileNameWithPath := path.Join(d.Path, ssTableContent[crc32Idx])
+		localCRC32, err := w.Client.RcloneCat(ctx, h.IP, localCRC32FileNameWithPath)
+		if err != nil {
+			return deduplicated, errors.Wrapf(err, "cannot get content of local CRC32 %s", localCRC32FileNameWithPath)
+		}
+
+		// If checksums are equal, then it means that SSTable can be deduplicated as there is no need in
+		// transferring it again to the remote storage.
+		// Deduplication here means to remove SSTable files from local storage.
+		if bytes.Equal(localCRC32, remoteCRC32) {
```

Comment:
AI explained the probability of crc32 hash collision of two files (two sstables data) this way:
```
For two files, the probability of a CRC32 hash collision is 1/2^32
 , which is approximately 2.3 × 10^-10 or 1 in 4.3 billion. This makes CRC32 quite reliable for small-scale use, but for very large datasets or critical applications, stronger hash functions (like SHA-256) are recommended to minimize collision risks.
```

Some external references https://arstechnica.com/civis/threads/crc32-hash-collision-probability.149670/
```
It's simple to figure out. No matter what your keys are the hash will be one of 2^32 different values. Assuming random input, the probability of any of these values appearing is equal. So, the probability of collision between the hashes of two given files is 1 / 2^32. The probability of collisions between any of N given files is (N - 1) / 2^32. <BR><BR>This is of course assuming that CRC-32 is perfectly unbiased, which it isn't, but the approximation isn't a big deal unless you have good knowledge of the key distribution.
```

We shouldn't bother.

Reply:
It's 0.00000002328%

Reply:
Yes, but those calculations are made for a single pair of files.
Single table in a big cluster could consist of millions of files.
Cluster has many tables.
SM manages many clusters.
And we would be really sad if a crc32 collision happened at any step.

The good thing is that having a name collision (result of replacing a node by ID) is a rare thing.
Also, new sstables are generated using the UUID, so that also lowers the pool of files exposed to potential crc32 collisions.
Having those in mind, it's always better to stay on the safer side and decrease the probability of the crc32 collision - especially that we already have the size information, so why shouldn't we use it?

Reply:
Checking the size makes sense. Fixing.